lib/s3_inventory.rb

```diff
@@ -0,0 +1,192 @@
+require "aws-sdk-s3"
+require "csv"
+require "discourse_event"
+
+class S3Inventory
+
+  class StorageError < RuntimeError; end
+
+  ::DiscourseEvent.on(:site_setting_saved) do |site_setting|
+    name = site_setting.name.to_s
+    Jobs.enqueue(:update_s3_inventory) if name.include?("s3_inventory") || name == "s3_upload_bucket"
+  end
+
+  attr_reader :inventory_id, :csv_filename, :model
+
+  CSV_KEY_INDEX ||= 1.freeze
+  CSV_ETAG_INDEX ||= 2.freeze
+  INVENTORY_PREFIX ||= "inventory".freeze
+
+  def initialize(s3_helper, type)
+    @s3_helper = s3_helper
+
+    if type == :upload
+      @inventory_id = "uploads"
+      @model = Upload
+    elsif type == :optimized
+      @inventory_id = "optimized"
+      @model = OptimizedImage
+    end
+  end
+
+  def file
+    @file ||= unsorted_files.sort_by { |file| -file.last_modified.to_i }.first
+  end
+
+  def list_missing
+    if file.blank?
+      Rails.logger.warn("Failed to list inventory from S3")
+      raise StorageError
+    end
+
+    DistributedMutex.synchronize("s3_inventory_list_missing_#{inventory_id}") do
+      current_db = RailsMultisite::ConnectionManagement.current_db
+      timestamp = Time.now.strftime("%Y-%m-%d-%H%M%S")
+      @tmp_directory = File.join(Rails.root, "tmp", INVENTORY_PREFIX, current_db, timestamp)
+      @archive_filename = File.join(@tmp_directory, File.basename(file.key))
+      @csv_filename = @archive_filename[0...-3]
+
+      FileUtils.mkdir_p(@tmp_directory)
+      copy_archive_to_tmp_directory
+      unzip_archive
+
+      begin
+        table_name = "#{inventory_id}_inventory"
+        connection = ActiveRecord::Base.connection.raw_connection
+        connection.exec("CREATE TEMP TABLE #{table_name}(key text UNIQUE, etag text PRIMARY KEY)")
```

Comment:
Do we need an index on the `etag` column?

Reply:
I thought `PRIMARY KEY` itself an index. no?